wagtail/models/sites.py

```diff
@@ -215,55 +254,75 @@ def get_site_root_paths():
+        return value will be cached for the current thread.
+        """
+        caching_enabled = getattr(settings, "WAGTAIL_PER_THREAD_SITE_CACHING", True)
+        if caching_enabled:
+            cached_result = cls._get_cached_site_root_paths()
+            if cached_result is not None:
+                return deepcopy(cached_result)
+
+        result = []
+        for site in Site.objects.get_all():
+            if getattr(settings, "WAGTAIL_I18N_ENABLED", False):
+                result.extend(
+                    [
                         SiteRootPath(
                             site.id,
-                            site.root_page.url_path,
+                            root_page.url_path,
                             site.root_url,
-                            site.root_page.locale.language_code,
+                            root_page.locale.language_code,
                         )
+                        for root_page in site.root_page.get_translations(
+                            inclusive=True
+                        ).select_related("locale")
+                    ]
+                )
+            else:
+                result.append(
+                    SiteRootPath(
+                        site.id,
+                        site.root_page.url_path,
+                        site.root_url,
+                        site.root_page.locale.language_code,
                     )
+                )
 
-            cache.set(
-                SITE_ROOT_PATHS_CACHE_KEY,
-                result,
-                3600,
-                version=SITE_ROOT_PATHS_CACHE_VERSION,
-            )
-
-        else:
-            # Convert the cache result to a list of SiteRootPath tuples, as some
-            # cache backends (e.g. Redis) don't support named tuples.
-            result = [SiteRootPath(*result) for result in result]
+        # cache result for the next request
+        if caching_enabled:
+            _site_root_paths_cache.value = result
 
         return result
 
     @staticmethod
-    def clear_site_root_paths_cache():
-        cache.delete(SITE_ROOT_PATHS_CACHE_KEY, version=SITE_ROOT_PATHS_CACHE_VERSION)
+    def _get_cached_site_root_paths() -> Union[List[SiteRootPath], None]:
+        return getattr(_site_root_paths_cache, "value", None)
+
+    @staticmethod
+    def clear_caches():
+        """
+        Convenience method for clearing both the `Site` and `SiteRootPath`
+        threadlocal caches.
+
+        As a courtesy, we clear cached data for threads being handled by
+        this VM/container when changes that might affect the return values
+        are detected.
+
+        Any threads running in other VMs/containers will continue to use
+        outdated cached data for the remaining life of the thread. But,
+        any new requests/threads (or active threads that have not yet queried
+        the cache) should pick up the updated values.
```

Comment:
Issue: This will result in different processes having different caches, leading to confusing and inconsistent behaviour. Processes may live for a very long time if there are no deployments, so it's important we not break functionality.

I think the only way to make this stable is to include some kind of TTL with the cache so processes get notified. Alternatively, something hierarchical, whereby the local cache has a very short TTL (a minute or so), and transparently hits a secondary cache that might be a little slower. Similar to L3 cache, RAM and disk behave.

Reply:
> This will result in different processes having different caches, leading to confusing and inconsistent behaviour.

I think you're ignoring the current state of play a little bit here. We already have different copies of the cached data everywhere, and we seem to be totally fine with that - That isn't the problem I'm solving here.

Reply:
> Processes may live for a very long time if there are no deployments, so it's important we not break functionality

Processes handled by the same container / VM will have their cached version of the data cleared should there be a change. Processes handled by other containers / VMs will not, but that isn't any worse than the current state of play.

Reply:
> transparently hits a secondary cache that might be a little slower.

Is this a suggestion to add cache backends into the mix on top of in-memory caching? That feels a tonne of extra complexity for not much gain here. From a pure system—design perspective, I get it. But realistically, the benefit would be minute compared solving the availability problem.